2001-01-05  Jan Hubicka  <jh@suse.cz>

	* reg-stack.c (swap_rtx_condition): Ensure that the transformation
	is valid.

2002-02-12  Gwenole Beauchesne  <gbeauchesne@mandrakesoft.com>

	* gcc.dg/20020212-5.c: New test.

--- gcc/testsuite/gcc.dg/20020122-5.c.jj	Wed Feb 27 15:41:30 2002
+++ gcc/testsuite/gcc.dg/20020122-5.c	Wed Feb 27 15:41:07 2002
@@ -0,0 +1,24 @@
+/* { dg-do run } */
+/* { dg-options "-O2" } */
+/* { dg-options "-O2 -mcpu=i686 -ffast-math" { target i?86-*-* } } */
+
+static int compare (double t1, double t2)
+{
+  if (t1 == t2)
+    return 0;
+  else if (t1 < t2)
+    return -1;
+  else
+    return +1;
+}
+
+int main (void)
+{
+  if (compare (1.0, 1.0) != 0)
+    abort ();
+  if (compare (3.0, 2.0) != 1)
+    abort ();
+  if (compare (2.0, 3.0) != -1)
+    abort ();
+  return 0;
+}
--- gcc/reg-stack.c.jj	2000/08/04 20:28:05	1.62
+++ gcc/reg-stack.c	2001/01/05 15:34:38	1.63
@@ -1261,7 +1261,17 @@ swap_rtx_condition (insn)
       pat = PATTERN (insn);
     }
 
-  return swap_rtx_condition_1 (pat);
+  if (swap_rtx_condition_1 (pat))
+    {
+      INSN_CODE (insn) = -1;
+      if (recog_memoized (insn) == -1)
+	{
+	  swap_rtx_condition_1 (pat);
+	  return 0;
+	}
+      return 1;
+    }
+  return 0;
 }
 
 /* Handle a comparison.  Special care needs to be taken to avoid
